### PR TITLE
Vigilante updated to use states

### PR DESCRIPTION
**Pull request by emt7553**

### CODE REVIEW

File: server/src/game/role/vigilante.rs [hunk 1]
```diff
@@ -30,43 +41,60 @@ impl RoleStateImpl for Vigilante {
...
+
+                            if target_ref.night_jailed(game){
+                                actor_ref.push_night_message(game, ChatMessage::TargetJailed);
+                                actor_ref.set_role_state(game, RoleState::Vigilante(self.clone()));
+                                return
+                            }
 
-                    let killed = target_ref.try_night_kill(actor_ref, game, GraveKiller::Role(Role::Vigilante), 1, false);
+                            let killed = target_ref.try_night_kill(actor_ref, game, GraveKiller::Role(Role::Vigilante), 1, false);
+                            self.state = VigilanteState::Loaded { bullets: bullets - 1 };
```
Comment (ItsSammyM) [inline server/src/game/role/vigilante.rs:70]: use saturating_sub here instead of -

File: server/tests/role.rs [hunk 2]
```diff
@@ -827,6 +827,24 @@ fn grave_contains_multiple_killers_roles() {
...
+        vigilante_retributionist: Vigilante,
+        vigilante_out_of_bullets: Vigilante,
+        retributionist: Retributionist
+
+    );
+    assert!(!vigilante_suicide.set_night_target(townie_b));
+    game.skip_to(PhaseType::Night, 2);
+    assert!(vigilante_suicide.set_night_target(townie_b))
+    //someone else can do this if they want 
+}
```
Comment (ItsSammyM) [inline server/tests/role.rs:847]: either make this test named properly and not have a ton of warnings and extra data, or delete it

File: client/src/game/messageListener.tsx [hunk 3]
```diff
@@ -276,6 +276,7 @@ export default function messageListener(packet: ToClientPacket){
                     GameScreen.instance?.closeMenu(ContentMenu.RoleSpecificMenu);
                 }
                 GAME_MANAGER.state.roleState = packet.roleState;
+                console.log(packet.roleState);
```
Comment (ItsSammyM) [inline client/src/game/messageListener.tsx:279]: remove this

File: client/src/game/roleState.d.tsx [hunk 4]
```diff
@@ -38,8 +38,7 @@ export type RoleState = {
     role: "trapper"
 } | {
     role: "vigilante",
-    bulletsRemaining: number,
-    willSuicide: boolean,
+    state: {type:"notLoaded"} | {type:"willSuicide"} | {type:"loaded",bullets:number} | {type:"suicided"}
```
Comment (ItsSammyM) [inline client/src/game/roleState.d.tsx:41]: jack doesn't want this to be in one line but I don't care

**Review by ItsSammyM — CHANGES_REQUESTED** [review]

**Review by Jack-Papel — APPROVED** [review]
LGTM But this PR made me enter my sad boy era

**Review by ItsSammyM — APPROVED** [review]

**Pull request merged**